File: dm-test/src/test/java/ro/johann/dm/test/api/service/decision/transfer/DecisionOutput.java
package ro.johann.dm.test.api.service.decision.transfer;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.Data;
import lombok.experimental.FieldDefaults;

import java.util.List;

import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL;
import static lombok.AccessLevel.PRIVATE;

@Data
@FieldDefaults(level = PRIVATE)
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(NON_NULL)
public class DecisionOutput {
  String id;
  String name;
  String description;
  String status;
  List<CriteriaOutput> criteria = List.of();
  List<AlternativeOutput> alternatives = List.of();
  List<PropertyOutput> properties = List.of();

  public void addAlternative(AlternativeOutput alternative) {
    alternatives.add(alternative);
  }

  public void addCriteria(CriteriaOutput criterion) {
    criteria.add(criterion);
  }
}

File: dm-decision-service/src/main/kotlin/ro/johann/dm/decision/api/transfer/DecisionOutput.kt
package ro.johann.dm.decision.api.transfer

import ro.johann.dm.decision.domain.Decision
import java.util.UUID

data class DecisionOutput(
  val id: UUID?,
  val name: String?,
  val description: String?,
  val criteria: List<CriteriaOutput> = emptyList(),
  val alternatives: List<AlternativeOutput> = emptyList(),
  val properties: List<PropertyOutput> = emptyList()
) {
  constructor(
    decision: Decision
  ) : this(
    id = decision.id,
    name = decision.name,
    description = decision.description,
    criteria = decision.criteria.map(::CriteriaOutput),
    alternatives = decision.alternatives.map(::AlternativeOutput),
    properties = decision.alternatives
      .asSequence()
      .flatMap { it.properties.asSequence() }
      .map(::PropertyOutput)
      .toList()
  )
}
